lsm/compaction_worker.go

```diff
@@ -0,0 +1,530 @@
+			continue
+		}
+		for c.started.Load() {
+			registrations, deRegistrations, err := c.processJob(job)
+			if err == nil {
+				regBatch := RegistrationBatch{
+					Compaction:      true,
+					JobID:           job.id,
+					Registrations:   registrations,
+					DeRegistrations: deRegistrations,
+				}
+				err = c.client.ApplyChanges(regBatch)
+				if err == nil {
+					// success
+					break
+				}
+			}
+			var tErr common.TektiteError
+			if errwrap.As(err, &tErr) {
+				if tErr.Code == common.Unavailable || tErr.Code == common.LevelManagerNotLeaderNode {
+					// transient unavailability - retry after delay
+					log.Debugf("transient error in compaction. will retry: %v", err)
+					time.Sleep(workerRetryInterval)
+					continue
+				}
+			}
+			log.Warnf("failed to process compaction:%v", err)
+			break
+		}
+	}
+}
+
+func (c *compactionWorker) processJob(job *CompactionJob) ([]RegistrationEntry, []RegistrationEntry, error) {
+	if job.isMove {
+		if len(job.tables) != 1 {
+			panic("move requires single run to move")
+		}
+		registrations, deRegistrations := c.moveTables(job.tables[0], job.levelFrom, job.preserveTombstones)
+		return registrations, deRegistrations, nil
+	}
+	start := time.Now()
+	tablesInMerge := 0
+	for _, overlapping := range job.tables {
+		tablesInMerge += len(overlapping)
+	}
+	var maxAddedTime uint64
+	tablesToMerge := make([][]tableToMerge, len(job.tables))
+	for i, overlapping := range job.tables {
+		tables := make([]tableToMerge, len(overlapping))
+		for j, t := range overlapping {
+			ssTable, err := c.cws.tableCache.GetSSTable(t.table.SSTableID)
+			if err != nil {
+				return nil, nil, err
+			}
+			if ssTable == nil {
+				return nil, nil, errwrap.Errorf("cannot process compaction job as cannot find sstable: %v (%s)", t.table.SSTableID,
+					string(t.table.SSTableID))
+			}
+			tables[j] = tableToMerge{
+				deadVersionRanges: t.table.DeadVersionRanges,
+				sst:               ssTable,
+				id:                t.table.SSTableID,
+			}
+			if t.table.AddedTime > maxAddedTime {
+				// We compute the maxAddedTime time - this is used for the AddedTime of any new sstables created.
+				maxAddedTime = t.table.AddedTime
+			}
+		}
+		tablesToMerge[i] = tables
+	}
+	mergeStart := time.Now()
+	var retProvider RetentionProvider
+	if c.cws.retentions {
+		retProvider = c
+	}
+	infos, err := mergeSSTables(common.DataFormatV1, tablesToMerge, job.preserveTombstones,
+		c.cws.cfg.CompactionMaxSSTableSize, job.lastFlushedVersion, job.id, retProvider, job.serverTime)
+	if err != nil {
+		return nil, nil, err
+	}
+	mergeDur := time.Now().Sub(mergeStart)
+	log.Debugf("merge for job %s took %d ms", job.id, mergeDur.Milliseconds())
+	// Now push the tables to the cloud store
+	var ids []sst.SSTableID
+	for _, info := range infos {
+		id := []byte(fmt.Sprintf("sst-%s", uuid.New().String()))
+		log.Debugf("compaction job %s created sstable %v", job.id, id)
+		ids = append(ids, id)
+		for {
```

Comment:
would it make sense to have a timeout here?

Reply:
The loop here only retries if there is some kind of unavailability of the object store - e.g. temporary network problem or S3 is down. Usually for these kinds of things we just want to keep retrying until it comes back up again. Consider what else we could do? If we timed out, after say, 10 minutes or a day, what then? The agent is useless if it cannot contact the object store, so  I guess we could terminate the process. But what if the network issue is resolved? Now we have restart the process. I think the best thing we can do in these kinds of situations is effectively have an infinite timeout.